### PR TITLE
Documentation for code

### LINKED ISSUE #22: Detailed documentation in code
- Follow Go Docs standards and add comments to every function.

### PR DESCRIPTION
**Pull request by balandi1**
- Added comments to every function, type and package.
- Merging of this pull request closes Issue 22.

### CODE REVIEW

**Review by pradykaushik — CHANGES_REQUESTED** [review]
README hasn't been updated.
For instance, there is nothing mentioned about the supported languages, libraries, etc. There are no instructions on how to add support for new libraries or languages and no instructions on how to write installation scripts.

**Comment by balandi1** [comment]
> README hasn't been updated.
> For instance, there is nothing mentioned about the supported languages, libraries, etc. There are no instructions on how to add support for new libraries or languages and no instructions on how to write installation scripts.

There is a separate issue for updating README and hence I will raise a separate PR for it. This PR just contains code comments and go docs.

File: builder/assignmentEnv.go [hunk 1]
```diff
@@ -1,3 +1,6 @@
+// Package builder implements routines to write dockerfile for assignment environment,
+// build its docker image and publish it to docker hub. It uses command pattern to
```
Comment (pradykaushik) [inline builder/assignmentEnv.go]: second sentence not required.

File: builder/assignmentEnv.go [hunk 2]
```diff
@@ -19,6 +22,9 @@ import (
 	"strings"
 )
 
+// baseEnvImage interface type represents functions for
+// performing different operations in order to build assignment
+// environment image.
 type baseEnvImage interface {
```
Comment (pradykaushik) [inline builder/assignmentEnv.go]: I don't see the need for this interface. Nowhere in the codebase is this interface type used.

File: builder/assignmentEnv.go [hunk 3]
```diff
@@ -28,15 +34,22 @@ type baseEnvImage interface {
 	undoBuild() error
 }
 
+// assignmentEnvironment struct type holds parameters required
```
Comment (pradykaushik) [inline builder/assignmentEnv.go]: maybe change the name to `assignmentEnvironmentImageBuilder`. With this the name of the corresponding source file will also have to change.

```go
// assignmentEnvironmentImageBuilder is used to build an image with an environment
// that satisfies all the language and library dependencies of an assignment.
//
// The image is built in phases to allow for rollback in situations where the execution of a phase fails.
// Once the configuration passes a verification phase, a Dockerfile is created and
// the docker client (here - https://github.com/moby/moby/tree/master/client) is used to generate the image.
// The image is then pushed to docker hub repository, if required.
```

**Review by pradykaushik — CHANGES_REQUESTED** [review]

File: builder/assignmentEnv.go (hunk 1 above)
Comment (balandi1) [reply builder/assignmentEnv.go]: Okay

File: builder/assignmentEnv.go (hunk 2 above)
Comment (balandi1) [reply builder/assignmentEnv.go]: Agreed. Removed it.

File: builder/assignmentEnv.go (hunk 3 above)
Comment (balandi1) [reply builder/assignmentEnv.go]: Okay

File: builder/assignmentEnvImageBuilder.go [hunk 4]
```diff
@@ -19,36 +21,41 @@ import (
...
+// The image is then pushed to docker hub repository, if required.
+type assignmentEnvironmentImageBuilder struct {
 	DockerfileInstructions bytes.Buffer
 	ImgBuildConfig         *imageBuildConfig
 	AssgnEnvConfig         *configurations.AssignmentEnvConfig
 	ImageExists            bool
 }
 
-type assignmentEnvironmentOption func(*assignmentEnvironment) error
+// assignmentEnvironmentOption is a function interface that is supplied as different
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: is function interface a thing?
Maybe rewrite the comment as shown below. You'll need to first rename the option to
`assignmentEnvironmentImageBuilderOption`.
```go
// assignmentEnvironmentImageBuilderOption represents options that can be used to help initialize
// an instance of assignmentEnvironmentImageBuilder.
// Each option is a closure that is responsible for initializing one or more members
// while instantiating assignmentEnvironmentImageBuilder.
```

File: builder/assignmentEnvImageBuilder.go [hunk 5]
```diff
@@ -19,36 +21,41 @@ import (
...
+// options while creating new instance of 'assignmentEnvironmentImageBuilder' type.
+// This function returns any error encountered.
+type assignmentEnvironmentOption func(*assignmentEnvironmentImageBuilder) error
 
-func newAssignmentEnvironment(options ...assignmentEnvironmentOption) (*assignmentEnvironment, error) {
-	assgnEnv := &assignmentEnvironment{}
+// newAssignmentEnvironment takes one or more options and returns new instance
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: rewrite this as shown below. You will also need to rename the constructor to `newAssignmentEnvironmentImageBuilder`.
```go
// newAssignmentEnvironmentImageBuilder constructs an instance of assignmentEnvironmentImageBuilder
// by applying each of the provided options.
// The construction of the object fails upon the failure of at least one of the given options.

File: builder/assignmentEnvImageBuilder.go [hunk 6]
```diff
@@ -19,36 +21,41 @@ import (
...
 	for _, opt := range options {
 		if err := opt(assgnEnv); err != nil {
-			return nil, errors.Wrap(err, "failed to create assignmentEnvironment instance")
+			return nil, errors.Wrap(err, "failed to create assignmentEnvironmentImageBuilder instance")
 		}
 	}
 	return assgnEnv, nil
 }
 
+// withImageBuildCfg takes imageBuildConfig as a parameter returns 'assignmentEnvironmentOption' function.
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: rewrite this as shown below.
```go
// withImageBuildCfg returns an assignmentEnvironmentImageBuilderOption for 
// initializing the image build config.
```

File: builder/assignmentEnvImageBuilder.go [hunk 7]
```diff
@@ -58,8 +65,10 @@ func withImageBuildCfg(imgBuildCfg *imageBuildConfig) assignmentEnvironmentOptio
 	}
 }
 
+// withAssgnEnvConfig takes AssignmentEnvConfig as a parameter returns 'assignmentEnvironmentOption' function.
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: rewrite this as shown below. Rename this option to `withAsgmtEnvConfig`. Refactor the other abbreviations of "assignment" in the code-base to "asgmt".
```go
// withAsgmtEnvConfig returns an assignmentEnvironmentImageBuilderOption for initializing
// the assignment environment configuration.
```

File: builder/assignmentEnvImageBuilder.go [hunk 8]
```diff
@@ -69,7 +78,10 @@ func withAssgnEnvConfig(assignCfgs *configurations.AssignmentEnvConfig) assignme
 	}
 }
 
-func (assgnEnv *assignmentEnvironment) verifyAndWriteInstructions() error {
+// verifyAndWriteInstructions checks whether the provided language image is already
+// present in docker hub and accordingly writes the Dockerfile from either base image
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: from either **the** base image.

File: builder/assignmentEnvImageBuilder.go [hunk 9]
```diff
@@ -69,7 +78,10 @@ func withAssgnEnvConfig(assignCfgs *configurations.AssignmentEnvConfig) assignme
...
+// or from the existing language image. It returns any error encountered.
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: don't have to mention that it returns an error.

File: builder/assignmentEnvImageBuilder.go [hunk 10]
```diff
@@ -69,7 +78,10 @@ func withAssgnEnvConfig(assignCfgs *configurations.AssignmentEnvConfig) assignme
 	}
 }
 
-func (assgnEnv *assignmentEnvironment) verifyAndWriteInstructions() error {
+// verifyAndWriteInstructions checks whether the provided language image is already
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: no language image is provided to this function.

File: builder/assignmentEnvImageBuilder.go [hunk 11]
```diff
@@ -88,7 +100,10 @@ func (assgnEnv *assignmentEnvironment) verifyAndWriteInstructions() error {
 	return nil
 }
 
-func (assgnEnv *assignmentEnvironment) verifyLanguage() error {
+// verifyLanguage searches the docker image for the given language of docker hub.
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: previously used "language image" and now using "docker image for the given language".
It is better to stick to one usage. I prefer "docker image for  ...".

File: builder/assignmentEnvImageBuilder.go [hunk 12]
```diff
@@ -88,7 +100,10 @@ func (assgnEnv *assignmentEnvironment) verifyAndWriteInstructions() error {
...
+// It uses ImageSearch function of docker client (here - https://github.com/moby/moby/tree/master/client)
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: this is implementation detail. Not required.

File: builder/assignmentEnvImageBuilder.go [hunk 13]
```diff
@@ -88,7 +100,10 @@ func (assgnEnv *assignmentEnvironment) verifyAndWriteInstructions() error {
...
+// and returns error if image is not found.
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: no need to mention that an error is returned.

File: builder/assignmentEnvImageBuilder.go [hunk 14]
```diff
@@ -119,7 +134,10 @@ func (assgnEnv *assignmentEnvironment) verifyLanguage() error {
 	return nil
 }
 
-func (assgnEnv *assignmentEnvironment) writeFromBaseImage() {
+// writeFromBaseImage writes the docker instructions to bytes buffer
+// in 'assignmentEnvironmentImageBuilder' instance staring from the base code runner image.
+// Which is then followed by the required language and its dependencies.
+func (assgnEnv *assignmentEnvironmentImageBuilder) writeFromBaseImage() {
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: what is the rationale behind naming the functions `writeFromXXX()`?

File: builder/assignmentEnvImageBuilder.go [hunk 15]
```diff
@@ -119,7 +134,10 @@ func (assgnEnv *assignmentEnvironment) verifyLanguage() error {
 	return nil
 }
 
-func (assgnEnv *assignmentEnvironment) writeFromBaseImage() {
+// writeFromBaseImage writes the docker instructions to bytes buffer
+// in 'assignmentEnvironmentImageBuilder' instance staring from the base code runner image.
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: you can just say **// writeFromBaseImage writes the docker instructions to a buffer.**

File: builder/assignmentEnvImageBuilder.go [hunk 16]
```diff
@@ -134,7 +152,10 @@ func (assgnEnv *assignmentEnvironment) writeFromBaseImage() {
 	assgnEnv.DockerfileInstructions.WriteString(strings.Join(instructions, "\n"))
 }
 
-func (assgnEnv *assignmentEnvironment) writeFromDependencies() {
+// writeFromDependencies writes the docker instructions to bytes buffer
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: follow similar guideline as shown above.

File: builder/assignmentEnvImageBuilder.go [hunk 17]
```diff
@@ -134,7 +152,10 @@ func (assgnEnv *assignmentEnvironment) writeFromBaseImage() {
...
+// in 'assignmentEnvironmentImageBuilder' instance staring from the respective language image.
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: ~staring~ starting.

File: builder/assignmentEnvImageBuilder.go [hunk 18]
```diff
@@ -156,7 +177,9 @@ func (assgnEnv *assignmentEnvironment) writeFromDependencies() {
 	assgnEnv.DockerfileInstructions.WriteString(strings.Join(instructions, "\n"))
 }
 
-func (assgnEnv *assignmentEnvironment) writeToDockerfile() error {
+// writeToDockerfile creates a Dockerfile at the specified location and writes
+// the instructions bytes buffer to it. It returns any error encountered.
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: no need to mention the data structure used. The documentation should be conceptual. It should include implementation detail only if something isn't obvious straight away or there was a specific message that is to be conveyed to the reader.

File: builder/assignmentEnvImageBuilder.go [hunk 19]
```diff
@@ -177,7 +200,11 @@ func (assgnEnv *assignmentEnvironment) writeToDockerfile() error {
 	return nil
 }
 
-func (assgnEnv *assignmentEnvironment) build() error {
+// build builds a docker image for the given language using the ImageBuild function
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: you can just say **// build a docker image ...**
Again implementation detail issue.

File: builder/assignmentEnvImageBuilder.go [hunk 20]
```diff
@@ -177,7 +200,11 @@ func (assgnEnv *assignmentEnvironment) writeToDockerfile() error {
...
+// of docker client (here - https://github.com/moby/moby/tree/master/client).
+// If the image is already present, then it simply pulls the image.
+// It returns any error encountered.
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: this is not required.

File: builder/assignmentEnvImageBuilder.go [hunk 21]
```diff
@@ -224,14 +251,20 @@ func (assgnEnv *assignmentEnvironment) build() error {
 	}
 }
 
-func (assgnEnv *assignmentEnvironment) publish() error {
+// publish pushes the built image to docker hub if it is not already present
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: **// publish the built image ...**

rewrite this comment as shown below.
```go
// publish the built image to docker hub, if required, if it is not already present.
```
Also, why not name the function publishImage(...) similar to pullImage(...)?

File: builder/assignmentEnvImageBuilder.go [hunk 22]
```diff
@@ -270,7 +303,11 @@ func (assgnEnv *assignmentEnvironment) pullImage() error {
 	return nil
 
 }
-func (assgnEnv *assignmentEnvironment) pushImage() error {
+
+// pushImage pushes image to docker hub using the ImagePush function
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: again implementation detail issue.

File: builder/assignmentEnvImageBuilder.go [hunk 23]
```diff
@@ -224,14 +251,20 @@ func (assgnEnv *assignmentEnvironment) build() error {
...
+// push operation.
+func (assgnEnv *assignmentEnvironmentImageBuilder) publish() error {
 	if !assgnEnv.ImageExists && assgnEnv.ImgBuildConfig.publishImage {
 		return assgnEnv.pushImage()
 	}
 	return nil
 }
 
-func (assgnEnv *assignmentEnvironment) pullImage() error {
+// pullImage pulls the required language image from docker hub using the ImagePull function
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: implementation detail issue.

File: builder/assignmentEnvImageBuilder.go [hunk 24]
```diff
@@ -224,14 +251,20 @@ func (assgnEnv *assignmentEnvironment) build() error {
...
+// of docker client (here - https://github.com/moby/moby/tree/master/client).
+// It returns any error encountered.
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: not required to be mentioned.

File: builder/assignmentEnvImageBuilder.go [hunk 25]
```diff
@@ -270,7 +303,11 @@ func (assgnEnv *assignmentEnvironment) pullImage() error {
 	return nil
 
 }
-func (assgnEnv *assignmentEnvironment) pushImage() error {
+
+// pushImage pushes image to docker hub using the ImagePush function
+// of docker client (here - https://github.com/moby/moby/tree/master/client).
+// It returns any error encountered.
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: not required to be mentioned.

**Review by pradykaushik — CHANGES_REQUESTED** [review]

File: builder/assignmentEnvImageBuilder.go (hunk 4 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Okay.

File: builder/assignmentEnvImageBuilder.go (hunk 5 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Done

File: builder/assignmentEnvImageBuilder.go (hunk 6 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Done

File: builder/assignmentEnvImageBuilder.go (hunk 7 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Done

File: builder/assignmentEnvImageBuilder.go (hunk 8 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Done

File: builder/assignmentEnvImageBuilder.go (hunk 9 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Okay

File: builder/assignmentEnvImageBuilder.go (hunk 10 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Yes. Rewrote the comment.

File: builder/assignmentEnvImageBuilder.go (hunk 11 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Changed

File: builder/assignmentEnvImageBuilder.go (hunk 12 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Done

File: builder/assignmentEnvImageBuilder.go (hunk 13 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Done

File: builder/assignmentEnvImageBuilder.go (hunk 14 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Depending on whether a docker image for given language is present on docker hub, the docker instructions are written either starting from base image or directly from dependencies. Hence two such functions.

File: builder/assignmentEnvImageBuilder.go (hunk 15 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Done

File: builder/assignmentEnvImageBuilder.go (hunk 16 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Done

File: builder/assignmentEnvImageBuilder.go (hunk 17 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Fixed

File: builder/assignmentEnvImageBuilder.go (hunk 18 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Okay

File: builder/assignmentEnvImageBuilder.go (hunk 19 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Done

File: builder/assignmentEnvImageBuilder.go (hunk 20 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Done

File: builder/assignmentEnvImageBuilder.go (hunk 21 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Done. Renamed publish() to publishImage(). Also got rid of pushImage(), it was not really needed.

File: builder/assignmentEnvImageBuilder.go (hunk 22 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Done

File: builder/assignmentEnvImageBuilder.go (hunk 23 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Done

File: builder/assignmentEnvImageBuilder.go (hunk 24 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Done

File: builder/assignmentEnvImageBuilder.go (hunk 25 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Done

File: builder/assignmentEnvImageBuilder.go [hunk 26]
```diff
@@ -0,0 +1,364 @@
...
+			if err != nil {
+				log.Println(err)
+				return
+			}
+		}()
+	}
+	return nil
+}
+
+// pullImage pulls the required docker image for given language from docker hub.
```
Comment (pradykaushik) [inline builder/assignmentEnvImageBuilder.go]: again is this pulling the image for the language or for the entire assignment environment?

Comment (pradykaushik) [reply builder/assignmentEnvImageBuilder.go]: Based on the answer, update the comment for build().

**Review by pradykaushik — APPROVED** [review]
Overall LGTM!
Just a couple of nit picks and then it's good to go.

File: builder/assignmentEnvImageBuilder.go (hunk 26 above)
Comment (balandi1) [reply builder/assignmentEnvImageBuilder.go]: Yes. For assignment environment

**Pull request merged**